inspirehep/modules/workflows/utils/grobid_authors_parser.py

```diff
@@ -0,0 +1,139 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2020 CERN.
+#
+# INSPIRE is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+from __future__ import absolute_import, division, print_function
+
+from inspire_schemas.builders import LiteratureBuilder
+from parsel import Selector
+
+
+class GrobidAuthors(object):
+    def __init__(self, xml_text):
+        if isinstance(xml_text, str):
+            xml_text = xml_text.decode('utf-8')
+        self._xml = Selector(text=xml_text, type="xml")
+        self._xml.remove_namespaces()
+        self._parsed_authors = self._xml.xpath("//author")
+        self._builder = None
+
+    def __getitem__(self, item):
+        return GrobidAuthor(self._parsed_authors[item])
+
+    def __len__(self):
+        return len(self._parsed_authors)
+
+    def parse_one(self):
+        """yield parsed authors one by one"""
+        self._builder = LiteratureBuilder()
+        for author in self:
+            yield {
+                'author': self._builder.make_author(
+                    full_name=author.fullname,
+                    raw_affiliations=author.raw_affiliations,
+                    emails=author.emails,
+                ),
+                'parsed_affiliations': author.processed_affiliations
+            }
+
+    def parse_all(self):
+        """Returns all authors at once as a list"""
+        return list(self.parse_one())
+
+
+class GrobidAuthor(object):
+    def __init__(self, author_selector):
+        self._author = author_selector
+
+    def _extract(self, source, path, type=None, text=False):
+        if type:
+            path += "[@type='{type}']".format(type=type)
+        if text:
+            path += "/text()"
+            return source.xpath(path)
+        return source.xpath(path)
+
+    def _extract_string(self, source, path, type=None, join_char=u' ', clean=True):
```

Comment:
not using the instance data (`self` is needed to access `_extract` only), so better make this a `@classmethod` instead). Also, isn't this just `join_char.join(_extract_strings_list(...))`?

Reply:
It's more like: ` join_char.join([x.strip() for x in _extract_strings_list(...)])` It's cleaner to call this method 6 times in a row (when building address) than calling original method in a join call with strip in a loop :P .